### PR TITLE
api: add 'badges' endpoint to project object

### PR DESCRIPTION
**Pull request by mwasilew**
'badges' endpoint will help rendering project badges with services like
shields.io. It currently has a problem with dynamically chaning colours
but otherwise works. Implementation details and interaction with
shields.io are described in the docs.

Signed-off-by: Milosz Wasilewski <milosz.wasilewski@linaro.org>

### CODE REVIEW

**Comment by mwasilew** [comment]
This isn't ideal as shields.io doesn't offer colour changes on dynamic badges. According to comments in shields.io tracker, this ticket is what we actually need: https://github.com/badges/shields/pull/1525. This will require slight changes in the SQUAD REST API but these changes are pretty trivial.

**Comment by mwasilew** [comment]
Failing check here: https://travis-ci.org/Linaro/squad/jobs/397292746 has nothing to do with the patch. Sth wrong with version of PyYaml in travis. The tests pass locally. Same issue on different reviews.

**Comment by danrue** [comment]
This is great! A few thoughts:

- No documentation on the /<group>/<project>/badge usage. It would be nice to not have the external dependency (and extra round trip too).
- I was really surprised that >90% is "green". In LKFT, that means that over 1000 things could fail and we'd still be green. Should either be configurable, or any failure >0 is not green. Also, I really care about eliminating failures - skips can be ignored for calculation purposes. Specifically, I'm suggesting:
  -  "green (#5cb85c) for pass rate above 90%" -> "green (#5cb85c) for fail rate of 0%"
  -  "orange (#f0ad4e) for pass rate between 50% and 90%" -> "orange (#f0ad4e) for fail rate between 0% and 10%"
  -  "red (#d9534f) for pass rate below 50%" -> "red (#d9534f) for fail rate above 10%"

screenshot from 2018-07-05 16-15-11

I tried using shields.io but had a bit of trouble. Do you have a working example I can try?

**Comment by mwasilew** [comment]
There are 2 options here:
 - use shields.io (which isn't perfect and doesn't support dynamic colours)
 - use local SVG badges (no dependency as far as I understand)
I'll leave the API endpoint as it can be used in some other use case and makes no harm.

This is what I'm getting with some old results from mainline
badge_orange

Configurable thresholds would probably be best, but I don't know how to do that conveniently. Any ideas?

If we go with your proposal, what happens with fail rate > 0 and pass rate > 90%? This case isn't covered (or I'm misreading the conditions).

**Comment by mwasilew** [comment]
I checked the appveyor tests in shields.io. The logic is pretty simple there:
 - # of passed = # of total tests -> badge is bright green
 - # of failed = 0 -> badge is green
 - # of passed = 0 -> badge is red
 - any case in between -> badge is orange
https://github.com/badges/shields/blob/master/server.js#L772

I'm fine with this logic. Any objections?

**Comment by danrue** [comment]
I based my suggested thresholds on failures instead of passes because otherwise you have to worry about the number of 'skips' (and in the near future, known issues) reported. Really, 0 failures should be the goal of any CI, so that should be what determines green. So, I agree with your proposal with regard to green.

For orange vs red, it seems too extreme to say that you only get red if there are 0 passes. At least on the projects i'm looking at, I would personally set 'red' at something like 10 failures (not 10%; 10). Perhaps it needs to be configurable?

**Comment by mwasilew** [comment]
Let's go with 0 passes for red for now. I'll try to make it configurable.

**Comment by mwasilew** [comment]
OK, I pushed the changes as described.

File: squad/api/rest.py [hunk 1]
```diff
@@ -112,6 +112,39 @@ class Meta:
...
+             project_status.tests_fail,
+             project_status.tests_skip)
+        pass_rate = 100 * float(project_status.tests_pass) / \
+            float(project_status.tests_pass + project_status.tests_fail + project_status.tests_skip)
+        output["project_pass_rate"] = "%0.2f" % pass_rate
+        # green for pass rate higher than 90%
+        # yellow for pass rate between 50% and 90%
+        # red for pass rate below 50%
+        colour = "f0ad4e"
+        if if project_status.tests_fail == 0:
```
Comment (terceiro) [inline squad/api/rest.py]: syntax error

**Comment by mwasilew** [comment]
sorry, I didn't run tests. Let's remove the API then. I'm not very attached to it :)

**Comment by mwasilew** [comment]
I added the docs for the standalone view and removed the API endpoint.

**Pull request merged**